File: fundiapp/lib/models/technician.dart
// lib/models/technician.dart

import 'package:netsaf_fund_app/models/portfolio.dart';

/// Handles the pivot price data from backend
class TechnicianServicePrice {
  final int id;
  final String name;
  final double minPrice;
  final double maxPrice;

  TechnicianServicePrice({
    required this.id,
    required this.name,
    required this.minPrice,
    required this.maxPrice,
  });

  factory TechnicianServicePrice.fromJson(Map<String, dynamic> json) {
    return TechnicianServicePrice(
      id: json['id'] ?? 0,
      name: json['name'] ?? '',
      minPrice: double.tryParse(json['pivot']?['min_price']?.toString() ?? '0') ?? 0.0,
      maxPrice: double.tryParse(json['pivot']?['max_price']?.toString() ?? '0') ?? 0.0,
    );
  }
}

class Technician {
  final int id;
  final String name;
  final String? phone;
  final String? profilePhoto;
  final String? nida;
  final String? idDocumentType;
  final String? idDocumentImage;
  final bool verified;
  final String? verificationStatus;
  final int registrationStep;
  final bool registrationCompleted;
  final String? bio;
  final int experience;
  final double rating;
  final double? hourlyRate;
  final String? area;
  final double? latitude;   // ✅ ADDED
  final double? longitude;  // ✅ ADDED
  final bool isOnline;
  final List<String> services;
  final List<TechnicianServicePrice> servicePrices;
  final List<PortfolioItem> portfolios;

  Technician({
    required this.id,
    required this.name,
    this.phone,
    this.profilePhoto,
    this.nida,
    this.idDocumentType,
    this.idDocumentImage,
    this.verified = false,
    this.verificationStatus,
    this.registrationStep = 0,
    this.registrationCompleted = false,
    this.bio,
    this.experience = 0,
    this.rating = 0.0,
    this.hourlyRate,
    this.area,
    this.latitude,          // ✅ ADDED
    this.longitude,         // ✅ ADDED
    this.isOnline = false,
    this.services = const [],
    this.servicePrices = const [],
    this.portfolios = const [],
  });

  factory Technician.fromJson(Map<String, dynamic> json, {bool isDetail = false}) {
    String name = json['name'] ?? '';
    if (name.isEmpty && json['user'] != null && json['user']['name'] != null) {
      name = json['user']['name'];
    }

    String? phone = json['phone'];
    if (phone == null && json['user'] != null && json['user']['phone'] != null) {
      phone = json['user']['phone'];
    }

    String? profilePhoto = json['profile_photo'];
    if (profilePhoto == null &&
        json['user'] != null &&
        json['user']['profile_photo'] != null) {
      profilePhoto = json['user']['profile_photo'];
    }

    String? nida = json['nida'];

    int parseExperience(dynamic value) {
      if (value == null) return 0;
      if (value is int) return value;
      if (value is String) return int.tryParse(value) ?? 0;
      return 0;
    }

    double parseDouble(dynamic value) {
      if (value == null) return 0.0;
      if (value is double) return value;
      if (value is int) return value.toDouble();
      if (value is String) return double.tryParse(value) ?? 0.0;
      return 0.0;
    }

    bool parseBool(dynamic value) {
      if (value == null) return false;
      if (value is bool) return value;
      if (value is int) return value == 1;
      if (value is String) {
        return value.toLowerCase() == 'true' || value == '1';
      }
      return false;
    }

    List<String> serviceNames = [];
    List<TechnicianServicePrice> servicePrices = [];

    if (json['service_prices'] != null && json['service_prices'] is List) {
      servicePrices = (json['service_prices'] as List)
          .map((e) => TechnicianServicePrice.fromJson(e as Map<String, dynamic>))
          .toList();
      serviceNames = servicePrices.map((s) => s.name).toList();
    } else if (json['services'] != null && json['services'] is List) {
      final raw = json['services'] as List;
      for (var item in raw) {
        if (item is String) {
          serviceNames.add(item);
        }
      }
    }

    return Technician(
      id: json['id'] ?? 0,
      name: name,
      phone: phone,
      profilePhoto: profilePhoto,
      nida: nida,
      idDocumentType: json['id_document_type'],
      idDocumentImage: json['id_document_image'],
      verified: parseBool(json['verified']),
      verificationStatus: json['verification_status'],
      registrationStep: parseExperience(json['registration_step']),
      registrationCompleted: parseBool(json['registration_completed']),
      bio: json['bio'],
      experience: parseExperience(json['experience']),
      rating: parseDouble(json['rating']),
      hourlyRate: json['hourly_rate'] != null
          ? parseDouble(json['hourly_rate'])
          : null,
      area: json['area'],
      // ✅ ADDED – parse latitude & longitude
      latitude: json['latitude'] != null ? parseDouble(json['latitude']) : null,
      longitude: json['longitude'] != null ? parseDouble(json['longitude']) : null,
      isOnline: parseBool(json['is_online']),
      services: serviceNames,
      servicePrices: servicePrices,
      portfolios: isDetail && json['portfolios'] != null
          ? (json['portfolios'] as List)
          .where((e) => e != null)
          .map((p) => PortfolioItem.fromJson(p))
          .toList()
          : [],
    );
  }
}

File: nearbyfundi_backend/app/Models/Technician.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Carbon\Carbon;

class Technician extends Model
{
    protected $fillable = [
        'user_id', 'profile_photo', 'bio', 'nida', 'experience', 'rating',
        'latitude', 'longitude', 'area', 'verified',
        'last_activity_at', 'is_online', 'hourly_rate',
        'location_updated_at',
        // FOR 4-STEP REGISTRATION
        'verification_status',
        'id_document_type',
        'id_document_image',
        'completed_jobs_count',
        'registration_step',
        'registration_completed',
    ];

    protected $casts = [
        'verified'            => 'boolean',
        'rating'              => 'float',
        'is_online'           => 'boolean',
        'last_activity_at'    => 'datetime',
        'location_updated_at' => 'datetime',
        'hourly_rate'         => 'decimal:2',
        'completed_jobs_count'=> 'integer',
        'registration_completed' => 'boolean',
    ];

    protected $hidden = [
        'nida', // sensitive — hidden from public JSON
    ];

    // ─── Relationships ──────────────────────────────────────────────
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function services()
    {
        return $this->belongsToMany(Service::class, 'technician_services');
    }

    /**
     * Service prices pivot with min/max
     */
    public function servicePrices()
    {
        return $this->belongsToMany(Service::class, 'technician_services')
                    ->withPivot('min_price', 'max_price')
                    ->withTimestamps();
    }

    public function portfolios()
    {
        return $this->hasMany(Portfolio::class);
    }

    public function posts()
    {
        return $this->hasMany(Post::class);
    }

    public function requests()
    {
        return $this->hasMany(ServiceRequest::class);
    }

    public function reviews()
    {
        return $this->hasMany(Review::class);
    }

    public function locationHistory()
    {
        return $this->hasMany(TechnicianLocationHistory::class);
    }

    // ─── Heartbeat (kept — used by nearby search / monitoring) ──────
    // NOTE: getComputedOnlineAttribute() + ONLINE_THRESHOLD_MINUTES
    // removed — they were never appended to JSON output, so nothing
    // in the app was actually reading them. Raw `is_online` still
    // drives the profile screen's badge (now removed client-side)
    // and remains available for admin monitoring / nearby search,
    // which likely depend on it directly.
    public function recordHeartbeat(?float $lat = null, ?float $lng = null): void
    {
        $this->is_online        = true;
        $this->last_activity_at = now();

        if ($lat !== null && $lng !== null) {
            $this->latitude            = $lat;
            $this->longitude           = $lng;
            $this->location_updated_at = now();
        }

        $this->save();
    }

    // ─── Rating Recalculation ──────────────────────────────────────
    public function recalculateRating(): void
    {
        $avg = $this->reviews()->avg('rating') ?? 0.0;
        $this->rating = round($avg, 1);
        $this->save();
    }

    public function incrementCompletedJobs(): void
    {
        $this->increment('completed_jobs_count');
    }

    // ─── Accessors ──────────────────────────────────────────────────
    public function getDisplayNameAttribute(): string
    {
        return $this->user->name ?? 'Unknown Technician';
    }
}